### PR TITLE
improve coverage for mir.ndslice.slice

### PR DESCRIPTION
**Pull request by wilzbach**
Let's increase the coverage of ndslice to 100% :)

### CODE REVIEW

File: source/mir/ndslice/slice.d [hunk 1]
```diff
@@ -1144,17 +1144,15 @@ struct Slice(size_t _N, _Range)
         mixin(mathIndexStrideCode);
     }
 
+    static if (hasPtrBehavior!PureRange)
```
Comment (wilzbach) [inline source/mir/ndslice/slice.d]: afaict this constructor is only called with pointers?

File: source/mir/ndslice/slice.d [hunk 2]
```diff
@@ -2778,6 +2794,17 @@ pure nothrow unittest
...
 
+unittest
+{
+    int[] arr = [3, 4, 5, 6, 7, 8];
+    auto ptr = arr.ptrShell;
+    auto ptrCopy = ptr.save;
+    ptrCopy[2] = 1;
+    ptrCopy._range.popFront;
+    assert(ptr[0] == 3);
+    assert(ptrCopy[0] == 4);
```
Comment (wilzbach) [inline source/mir/ndslice/slice.d]: is this a correct test for the `save` behavior?

File: source/mir/ndslice/slice.d (hunk 1 above)
Comment (9il) [inline source/mir/ndslice/slice.d]: no

File: source/mir/ndslice/slice.d [hunk 3]
```diff
@@ -1144,17 +1144,15 @@ struct Slice(size_t _N, _Range)
...
         foreach (i; Iota!(0, PureN))
             _lengths[i] = lengths[i];
         foreach (i; Iota!(0, PureN))
             _strides[i] = strides[i];
-        static if (hasPtrBehavior!PureRange)
-            _ptr = range;
-        else
-            _ptr._range = range;
 
+        _ptr = range;
```
Comment (9il) [inline source/mir/ndslice/slice.d]: but this is ok

File: source/mir/ndslice/slice.d [hunk 4]
```diff
@@ -1720,6 +1718,24 @@ struct Slice(size_t _N, _Range)
         assert(slice(indexP) == 4);  // Math & Fortran order
     }
 
+
```
Comment (9il) [inline source/mir/ndslice/slice.d]: remove line please

File: source/mir/ndslice/slice.d [hunk 5]
```diff
@@ -1720,6 +1718,24 @@ struct Slice(size_t _N, _Range)
...
+    static if (doUnittest) unittest
```
Comment (9il) [inline source/mir/ndslice/slice.d]: please use different lines

File: source/mir/ndslice/slice.d (hunk 2 above)
Comment (9il) [inline source/mir/ndslice/slice.d]: no, see dummyranges module in phobos sources

Comment (9il) [inline source/mir/ndslice/slice.d]: You should define a range based on class representation

File: source/mir/ndslice/slice.d [hunk 6]
```diff
@@ -2869,6 +2893,24 @@ pure nothrow unittest
     assert(b[0] == 2);
 }
 
+unittest
+{
+    // test save behavior
```
Comment (9il) [inline source/mir/ndslice/slice.d]: the same, this test does not actually test save.

File: source/mir/ndslice/slice.d (hunk 3 above)
Comment (wilzbach) [inline source/mir/ndslice/slice.d]: doesn't work without the restriction

```
 cannot implicitly convert expression (range) of type Result to PtrShell!(Result)
```

Comment (9il) [inline source/mir/ndslice/slice.d]: oh, understand now. Please don' fix this constructor. This is for internal use only

Comment (wilzbach) [inline source/mir/ndslice/slice.d]: coverage says it is not used at all :/

Comment (9il) [inline source/mir/ndslice/slice.d]: This is just not covered, but this is used by package

File: source/mir/ndslice/slice.d [hunk 7]
```diff
@@ -2867,6 +2886,12 @@ pure nothrow unittest
     r = t.front;
     r.b = r.a * 2;
     assert(b[0] == 2);
+
+    // test save
+    auto c = new int[20];
+    auto tCopy = t.save;
+    tCopy.ptrs[0] = c.ptr;
+    assert(t.front.a == 1);
```
Comment (wilzbach) [inline source/mir/ndslice/slice.d]: using dummy ranges didn't work - see #163

File: source/mir/ndslice/slice.d (hunk 3 above)
Comment (wilzbach) [inline source/mir/ndslice/slice.d]: > This is just not covered, but this is used by package

where?

File: source/mir/ndslice/slice.d (hunk 7 above)
Comment (9il) [inline source/mir/ndslice/slice.d]: this is wrong test

File: source/mir/ndslice/slice.d (hunk 3 above)
Comment (9il) [inline source/mir/ndslice/slice.d]: > where?

The line with `_ptr` is used, so the constructor is used.

Comment (wilzbach) [inline source/mir/ndslice/slice.d]: maybe we are talking about different things, this constructor is always called with ranges that have `PtrBehavior`, so this line:

```
_ptr._range = range;
```

is never called.

Comment (9il) [inline source/mir/ndslice/slice.d]: This line is never called for current test, but may be called for user code

Comment (9il) [inline source/mir/ndslice/slice.d]: Anyway, this is correct line and i don't what this be removed.

File: source/mir/ndslice/slice.d [hunk 8]
```diff
@@ -1144,27 +1144,15 @@ struct Slice(size_t _N, _Range)
         mixin(mathIndexStrideCode);
     }
 
-    this(ref in size_t[PureN] lengths, ref in sizediff_t[PureN] strides, PureRange range)
+    this(S)(ref in size_t[PureN] lengths, ref in sizediff_t[PureN] strides, S range)
+    if(hasPtrBehavior!PureRange || is(PtrShell!PureRange == S))
```
Comment (wilzbach) [inline source/mir/ndslice/slice.d]: turns out that `_ptr` can only be either of them - see above:

```
    static if (hasPtrBehavior!PureRange)
        PureRange _ptr;
    else
        PtrShell!PureRange _ptr;
```

File: source/mir/ndslice/slice.d [hunk 9]
```diff
@@ -2778,6 +2790,18 @@ pure nothrow unittest
     }
 }
 
+unittest
+{
+    // test save with dynamic arrays
```
Comment (wilzbach) [inline source/mir/ndslice/slice.d]: dynamic arrays are handled differently in contrast to ranges, thus the extra test.

File: source/mir/ndslice/slice.d (hunk 8 above)
Comment (9il) [inline source/mir/ndslice/slice.d]: Please stop breaking logic of this constructor

Comment (9il) [inline source/mir/ndslice/slice.d]: I need this logic in this module

Comment (9il) [inline source/mir/ndslice/slice.d]: I am already spent a lot of time because you want to cover 1 line

File: source/mir/ndslice/slice.d [hunk 10]
```diff
@@ -2810,8 +2834,6 @@ private template PtrTuple(Names...)
             foreach (i, ref ptr; ptrs)
                 static if (isPointer!(Ptrs[i]))
                     p.ptrs[i] = ptr;
-                else
-                    p.ptrs[i] = ptr.save;
```
Comment (9il) [inline source/mir/ndslice/slice.d:2814]: Why????

**Comment by 9il** [comment]
If you can not cover something, do not remove it. Let it be uncovered please

**Comment by wilzbach** [comment]
> I am already spent a lot of time because you want to cover 1 line

will rebase, s.t. that only the math test is in there and throw away the other stuff.

**Comment by wilzbach** [comment]
> If you can not cover something, do not remove it. Let it be uncovered please

rebased.

**Pull request merged**